src/Azure/Orleans.Transactions.AzureStorage/TransactionalState/AzureTableTransactionalStateStorage.cs

```diff
@@ -0,0 +1,184 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.WindowsAzure.Storage.Table;
+using Newtonsoft.Json;
+using Orleans.Transactions.Abstractions;
+
+namespace Orleans.Transactions.AzureStorage
+{
+    public class AzureTableTransactionalStateStorage<TState> : ITransactionalStateStorage<TState>
+        where TState : class, new()
+    {
+        private readonly CloudTable table;
+        private readonly string partition;
+        JsonSerializerSettings JsonSettings;
+        private readonly ILogger logger;
+
+        private KeyEntity key;
+        private List<StateEntity> states;
+
+        public AzureTableTransactionalStateStorage(CloudTable table, string partition, JsonSerializerSettings JsonSettings, ILogger<AzureTableTransactionalStateStorage<TState>> logger)
+        {
+            this.table = table;
+            this.partition = partition;
+            this.JsonSettings = JsonSettings;
+            this.logger = logger;
+        }
+
+        public async Task<TransactionalStorageLoadResponse<TState>> Load()
+        {
+            try
+            {
+                Task<KeyEntity> keyTask = ReadKey();
+                Task<List<StateEntity>> statesTask = ReadStates();
+                this.key = await keyTask;
+                this.states = await statesTask;
+                if (string.IsNullOrEmpty(this.key.ETag))
+                {
+                    return new TransactionalStorageLoadResponse<TState>();
+                }
+                TState commitedState = TryFindState(this.key.CommittedTransactionId, out TState found) ? found : new TState();
+                var pendingStates = states.Select(s => new PendingTransactionState<TState>(s.TransactionId, s.SequenceId, s.GetState<TState>(this.JsonSettings))).ToList();
+                return new TransactionalStorageLoadResponse<TState>(this.key.ETag, commitedState, this.key.Metadata, pendingStates);
+            } catch(Exception ex)
+            {
+                this.logger.LogError("Transactional state load failed {Exception}.", ex);
+                throw;
+            }
+        }
+
+        public async Task<string> Persist(string expectedETag, string metadata, List<PendingTransactionState<TState>> statesToPrepare)
+        {
+            try
+            {
+                var batchOperation = new TableBatchOperation();
+
+                this.key.ETag = expectedETag;
+                this.key.Metadata = metadata;
+                if (string.IsNullOrEmpty(this.key.ETag))
+                    batchOperation.Insert(this.key);
+                else
+                    batchOperation.Replace(this.key);
```

Comment:
TableOperation.InsertOrReplace does this if for you

Reply:
Insert or replace will not respect etag.  We want to respect etag if it's set.